BT/wired split runtime transport selection (#2886)
zmkfirmware/zmk: docs/docs/features/split-keyboards.md

Bluetooth is the most well tested, and fully support transport available in ZMK. It supports topologies with multiple peripherals, dongle mode, and all data passing currently supported by ZMK.

This transport will be enabled for designs that set `CONFIG_ZMK_SPLIT=y` and have `CONFIG_ZMK_BLE=y` set by a supported MCU/controller.

### Full-Duplex Wired (UART)

The full-duplex wired UART transport allows for fully functional communicate between one central and one peripheral. Unlike the Bluetooth transport, topologies with multiple peripherals are _not_ yet supported by this transport.

This transport will be enabled for designs that set `CONFIG_ZMK_SPLIT=y` and have a configured devicetree including a node with `compatible = "zmk,wired-split";`.

:::note[Full Duplex vs Half Duplex]
Future single-wire, half-duplex UART support, which is planned, will allow using wired ZMK with designs like Corne, Sweep, etc. that use only a single GPIO pin for bidirectional communication between split sides.

Until half-duplex support is completed, those particular designs will not work with the wired split transport, and can only be used with the Bluetooth transport.
:::

### Runtime Switching

> For advanced designs, there is experimental support for switching between the two available transports. Supporting runtime switching is _not_ recommended for casual users or designers, as doing this right to avoid the damage from hot plugging cables is _non-trivial_ at best, and not implemented for classic designs like Corne, Sofle, etc. Attempting to use this functionality on those existing design _WILL_ damage your hardware. Currently, there are no open source/reference designs that implement this functionality, and only experienced designers with extensive EE knowledge should attempt to implement a design with this functionality.

Author: I don't personally have any interest in doing so. It's really hard to do right, and I'm hesitant to "make this accessible" for folks who might then miss something and then damage hardware.

Since peripherals communicate through centrals, the key and sensor events originating from them will naturally have a larger latency, especially with a wireless split communication protocol.
For the currently used BLE-based transport, split communication increases the average latency by 3.75ms with a worst case increase of 7.5ms.

## Split Transports

> ZMK supports two transports for connecting split pieces: Bluetooth and full-duplex wired UART. Only one transport can be active at a time, so designs involving some portions connected via Bluetooth and others via full-duplex wired are _not_ supported.

Reviewer: ```suggestion
ZMK supports two transports for connecting split parts: Bluetooth and full-duplex wired UART. Only one transport can be active at a time, so designs involving some portions connected via Bluetooth and others via full-duplex wired are _not_ supported.
```

A while ago we agreed on the terminology of "parts" as opposed to "pieces" to describe splits with 3+....well, parts.

:::warning[Hot Plugging Cables]

Many popular cables, in particular, TRRS/TRS cables, can cause irreparable damage to controllers if they are inserted or removed when power is already present on them. Whether or not you are using the wired split functionality or not, _never_ insert or remove such a cable when a controller is powered by USB _or_ battery.

:::

### Bluetooth

[Bluetooth](./bluetooth.md) is the most well tested and flexible transport available in ZMK. Using Bluetooth, a central can connect to multiple peripherals, enabling the use of a [dongle](../development/hardware-integration/dongle.mdx) to improve battery life, or allowing for multi-part split keyboards.

This transport will be enabled for designs that set `CONFIG_ZMK_SPLIT=y` and have `CONFIG_ZMK_BLE=y` set by a supported MCU/controller.

### Full-Duplex Wired (UART)

> The full-duplex wired UART transport is a recent addition, and is intended for testing by early adopters. It allows for fully functional communicate between one central and one peripheral. Unlike the Bluetooth transport, the central cannot currently have more than one peripheral connected via wired split.

Reviewer: ```suggestion
The full-duplex wired UART transport is a recent addition, and is intended for testing by early adopters. It allows for fully functional communication between one central and one peripheral. Unlike the Bluetooth transport, the central cannot currently have more than one peripheral connected via wired split.
```

> This transport will be enabled for designs that set `CONFIG_ZMK_SPLIT=y` and have a configured devicetree including a node with `compatible = "zmk,wired-split";`.

Reviewer: ```suggestion
This transport will be enabled for designs that set `CONFIG_ZMK_SPLIT=y` and have a node with `compatible = "zmk,wired-split";` present in their devicetree configuration.
```
